### PR TITLE
fix(builder): fix builder state machine replacement tracking

### PR DESCRIPTION
**Pull request by dancoombs**
Implement a better tracking mechanism for state machine replacement underpriced tracking.

TODO:

- [x] Unit tests for new tracking
- [x] Update docs with state machine diagram and description

### CODE REVIEW

File: crates/builder/src/task.rs [hunk 1]
```diff
@@ -91,8 +91,10 @@ pub struct Args {
     pub max_blocks_to_wait_for_mine: u64,
     /// Percentage to increase the fees by when replacing a bundle transaction
     pub replacement_fee_percent_increase: u64,
-    /// Maximum number of times to increase the fees when replacing a bundle transaction
-    pub max_fee_increases: u64,
+    ///
```
Comment (0xfourzerofour) [inline crates/builder/src/task.rs:94]: should there be comments here?

File: crates/builder/src/transaction_tracker.rs [hunk 2]
```diff
@@ -399,6 +433,12 @@ where
         let nonce = self.get_external_nonce().await.unwrap_or(self.nonce);
         self.set_nonce_and_clear_state(nonce);
     }
+
+    fn abandon(&mut self) {
+        self.has_abandoned = true;
```
Comment (0xfourzerofour) [inline crates/builder/src/transaction_tracker.rs:438]: Nit: has_abandoned sounds like something in the op is abandoned not the op itself, I think it should be called was_abandoned or is_abandoned

**Review by 0xfourzerofour — APPROVED** [review]
looks good just a nit

File: crates/builder/src/task.rs (hunk 1 above)
Comment (dancoombs) [reply crates/builder/src/task.rs:94]: Yup, done

File: crates/builder/src/transaction_tracker.rs (hunk 2 above)
Comment (dancoombs) [reply crates/builder/src/transaction_tracker.rs:438]: We already have `has_dropped` on this class, so I kept the same naming

**Review by 0xfourzerofour — APPROVED** [review]

**Pull request merged**